libcache/cache.c

```diff
@@ -0,0 +1,696 @@
+	void *data;
+	unsigned char flags;
+};
+
+
+typedef struct {
+	cacheline_t *timestamps;
+	cacheline_t *tags[LIBCACHE_NUM_WAYS];
+	cacheline_t lines[LIBCACHE_NUM_WAYS];
+	size_t count;
+} cacheset_t;
+
+
+struct cachectx_s {
+	cacheset_t *sets;
+
+	size_t srcMemSize;
+	size_t lineSize;
+	size_t numLines;
+	size_t numSets;
+
+	uint64_t tagMask;
+	uint64_t setMask;
+	uint64_t offMask;
+
+	uint8_t offBitsNum;
+	uint8_t setBitsNum;
+	uint8_t tagBitsNum;
+
+	cache_readCb_t readCb;
+	cache_writeCb_t writeCb;
+
+	handle_t lock;
+};
+
+
+static uint64_t cache_generateMask(int numBits)
+{
+	return ((uint64_t)1 << numBits) - 1;
+}
+
+
+cachectx_t *cache_init(size_t srcMemSize, size_t size, size_t lineSize, cache_writeCb_t writeCb, cache_readCb_t readCb)
+{
+	int err;
+	cachectx_t *cache = NULL;
+
+	if (srcMemSize == 0 || size == 0 || lineSize == 0) {
+		cache = NULL;
+	}
+	else if (size % lineSize != 0 || (size / lineSize) % LIBCACHE_NUM_WAYS != 0) {
+		cache = NULL;
+	}
+	else {
+		cache = malloc(sizeof(cachectx_t));
+
+		if (cache != NULL) {
+			cache->srcMemSize = srcMemSize;
+			cache->lineSize = lineSize;
+			cache->numLines = size / cache->lineSize;
+
+			cache->numSets = cache->numLines / LIBCACHE_NUM_WAYS;
+
+			cache->sets = calloc(cache->numSets, sizeof(cacheset_t));
+
+			if (cache->sets != NULL) {
+				cache->offBitsNum = LOG2(cache->lineSize);
+				cache->setBitsNum = LOG2(cache->numSets);
+				cache->tagBitsNum = LIBCACHE_ADDR_WIDTH - cache->setBitsNum - cache->offBitsNum;
+
+				cache->tagMask = cache_generateMask(cache->tagBitsNum);
+				cache->setMask = cache_generateMask(cache->setBitsNum);
+				cache->offMask = cache_generateMask(cache->offBitsNum);
+
+				cache->readCb = readCb;
+				cache->writeCb = writeCb;
+
+				err = mutexCreate(&cache->lock);
+				if (err < 0) {
+					free(cache->sets);
+					free(cache);
+					cache = NULL;
+				}
+			}
+			else {
+				free(cache);
+				cache = NULL;
+			}
+		}
+	}
+
+	return cache;
+}
+
+
+static uint64_t cache_computeAddr(const cachectx_t *cache, uint64_t tag, uint64_t setIndex)
+{
+	uint64_t temp = (tag << cache->setBitsNum) | setIndex;
+	return temp <<= cache->offBitsNum;
+}
+
+
+static ssize_t cache_flushLine(cachectx_t *cache, cacheline_t *linePtr, uint64_t addr)
+{
+	ssize_t written = 0, position = 0;
+	size_t left = cache->lineSize;
+
+	if (linePtr != NULL && IS_VALID(linePtr->flags) && IS_DIRTY(linePtr->flags)) {
+		while (left > 0) {
+			written = cache->writeCb(addr, (unsigned char *)linePtr->data + position, left);
+
+			if (written <= 0) {
+				position = -EIO;
+				break;
+			}
+
+			left -= written;
+			addr += written;
+			position += written;
+		}
+
+		if (position == (ssize_t)cache->lineSize) {
+			CLEAR_DIRTY(linePtr->flags);
+		}
+	}
+
+	return position;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j, ret = EOK, err;
+	uint64_t addr;
+	ssize_t written;
+	cacheline_t *linePtr;
+
+	if (cache == NULL) {
+		ret = -EINVAL;
+	}
```

Comment:
I wonder if this check is really needed. Why would someone pass NULL?

Reply:
I've realized that this check is in fact unnecessary as it doesn't stop the user from passing any other faulty pointer to the function anyway. Thank you, I've removed it from this and the other functions.